plutus-contract/test/Spec/TxConstraints/MustSpendScriptOutput.hs

```diff
@@ -0,0 +1,289 @@
+    ledgerTx1 <- submitTxConstraintsWith lookups1 tx1
+    awaitTxConfirmed $ Tx.getCardanoTxId ledgerTx1
+
+    scriptUtxos <- utxosAt mustSpendScriptOutputWithMatchingDatumAndValueScrAddress
+    let lookups2 = Constraints.plutusV1OtherScript mustSpendScriptOutputWithMatchingDatumAndValueVal <>
+            Constraints.unspentOutputs scriptUtxos
+        tx2 = Constraints.mustSpendScriptOutputWithMatchingDatumAndValue mustSpendScriptOutputWithMatchingDatumAndValueValHash (\d -> d == asDatum offChainMatchingDatum) (\v -> v == offChainMatchingValue) (asRedeemer [(onChainMatchingDatum, onChainMatchingValue)])
+    ledgerTx4 <- submitTxConstraintsWith @MustSpendScriptOutputValType lookups2 tx2
+    awaitTxConfirmed $ Tx.getCardanoTxId ledgerTx4
+
+trace :: Contract () Empty ContractError () -> Trace.EmulatorTrace ()
+trace contract = do
+    void $ Trace.activateContractWallet w1 contract
+    void $ Trace.waitNSlots 5
+
+-- | Uses onchain and offchain constraint mustSpendScriptOutput to spend all UtxOs locked by the script
+validUseOfMustSpendScriptOutputUsingAllScriptOutputs :: TestTree
+validUseOfMustSpendScriptOutputUsingAllScriptOutputs =
+    checkPredicateOptions
+    defaultCheckOptions
+    "Successful use of mustSpendScriptOutput for all script's UtxOs"
+    (valueAtAddress mustSpendScriptOutputScrAddress (== Ada.lovelaceValueOf 0)
+    .&&. assertValidatedTransactionCount 2)
+    (void $ trace $ mustSpendScriptOutputsContract 5 5)
+
+-- | Uses onchain and offchain constraint mustSpendScriptOutput to spend some of the UtxOs locked by the script
+validUseOfMustSpendScriptOutputUsingSomeScriptOutputs :: TestTree
+validUseOfMustSpendScriptOutputUsingSomeScriptOutputs =
+    checkPredicateOptions
+    defaultCheckOptions
+    "Successful use of mustSpendScriptOutput for some of the script's UtxOs"
+    (valueAtAddress mustSpendScriptOutputScrAddress (== Value.scale 2 utxoValue)
+    .&&. assertValidatedTransactionCount 2)
+    (void $ trace $ mustSpendScriptOutputsContract 5 3)
+
+-- | Contract error occurs when offchain mustSpendScriptOutput constraint is used with a UTxO belonging to the w1 pubkey address, not the script (not of type ScriptChainIndexTxOut).
+contractErrorWhenMustSpendScriptOutputUsesWrongTxoOutRef :: TestTree
+contractErrorWhenMustSpendScriptOutputUsesWrongTxoOutRef =
+    let contract :: Contract () Empty ContractError () = do
+            let lookups1 = Constraints.typedValidatorLookups typedMustSpendScriptOutputValidator
+                tx1 = mustPayToTheScriptWithMultipleOutputs 3 []
+            ledgerTx1 <- submitTxConstraintsWith lookups1 tx1
+            awaitTxConfirmed $ Tx.getCardanoTxId ledgerTx1
+
+            scriptUtxos <- utxosAt mustSpendScriptOutputScrAddress
+            w1Utxos <- utxosAt (mockWalletAddress w1)
+            let w1Utxo = fst $ M.elemAt 2 w1Utxos
+                lookups2 = Constraints.plutusV1OtherScript mustSpendScriptOutputVal <>
```

Comment:
You can use typedValidatorLookups here I think

Reply:
I'll leave one test's contract with `plutusV1OtherScript` an the rest with `typedValidatorLookups`